framework/src/partitioner/CopyMeshPartitioner.C

```diff
@@ -0,0 +1,218 @@
+//* This file is part of the MOOSE framework
+//* https://mooseframework.inl.gov
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CopyMeshPartitioner.h"
+
+#include "MooseApp.h"
+#include "MooseMesh.h"
+#include "MooseRandom.h"
+
+#include "libmesh/elem.h"
+
+// TIMPI includes
+#include "timpi/communicator.h"
+#include "timpi/parallel_sync.h"
+
+registerMooseObject("MooseApp", CopyMeshPartitioner);
+
+InputParameters
+CopyMeshPartitioner::validParams()
+{
+  InputParameters params = MoosePartitioner::validParams();
+  params.addClassDescription("Assigns element to match the partitioning of another mesh. If in a "
+                             "child application, defaults to the parent app mesh if the other mesh "
+                             "is not specified programmatically.");
+  params.addPrivateParam<MooseMesh *>("source_mesh");
+
+  return params;
+}
+
+CopyMeshPartitioner::CopyMeshPartitioner(const InputParameters & params) : MoosePartitioner(params)
+{
+  if (isParamValid("source_mesh"))
+    _base_mesh = getParam<MooseMesh *>("source_mesh");
+  else if (!_app.isUltimateMaster() && _app.masterMesh())
+    _base_mesh = _app.masterMesh();
+  else
+    mooseError("Expecting either a parent app with a mesh to copy the partitioning from, a "
+               "'source_mesh' (private) parameter to be set programmatically.");
+}
+
+std::unique_ptr<Partitioner>
+CopyMeshPartitioner::clone() const
+{
+  return _app.getFactory().clone(*this);
+}
+
+void
+CopyMeshPartitioner::_do_partition(MeshBase & mesh, const unsigned int /*n*/)
+{
+  mooseAssert(_base_mesh, "Should have a base mesh to copy the partitioning from");
+
+  // First check that it makes sense to copy the partitioning
+  // If we use the same number of procs to partition this mesh than to partition the source mesh, we
+  // are effectively matching regions to processes the same way in both meshes. Makes sense.
+  // If we use more procs, this will leave some processes with no elements. It is not ideal, let's
+  // just give a warning. This does not happen in practice with MultiApps.
+  // If we use less procs (N_source > N_current), we error. If we could avoid erroring, then either:
+  // - the elements on our mesh always get matched to elements from only N_current processes among
+  //   the N_source processes. We can accomodate that
+  // - the elements on our mesh get matched to more processes than we have. We would
+  //   need a heuristic (for example a nested partitioner) to re-distribute these elements. We won't
+  //   support it
+  std::set<unsigned int> pids_used;
+
+  // We cannot get the point locator only on some ranks of the base mesh, so we must error here
+  if (_base_mesh->comm().size() > mesh.comm().size())
```

Comment:
If we want to be able to see owned source mesh elements from the corresponding target mesh's owned elements after the partitioning is done, then checking the communicator sizes isn't enough - we also need each process to be the same rank() on each communicator.

Reply:
added an assert, not sure when this would happen

Reply:
Yeah, me either, but then I wouldn't have considered the case where the communicators are different sizes to begin with.